service/matching/taskReader.go

```diff
@@ -240,22 +231,23 @@ func (tr *taskReader) addTasksToBuffer(
 			tr.tlMgr.taskAckManager.setReadLevel(t.GetTaskId())
 			continue
 		}
-		if !tr.addSingleTaskToBuffer(t) {
-			return false // we are shutting down the task queue
+		if err := tr.addSingleTaskToBuffer(ctx, t); err != nil {
+			return err
 		}
 	}
-	return true
+	return nil
 }
 
 func (tr *taskReader) addSingleTaskToBuffer(
+	ctx context.Context,
 	task *persistencespb.AllocatedTaskInfo,
-) bool {
+) error {
 	tr.tlMgr.taskAckManager.addTask(task.GetTaskId())
 	select {
 	case tr.taskBuffer <- task:
-		return true
-	case <-tr.shutdownChan:
-		return false
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
```

Comment:
`context canceled`? It is really useful? Probably yes.

Reply:
What was really useful to me was for this function to return error. Could just as effectively be `errors.New("oops")` but `context canceled` is indeed the reason for the failure so it felt right.